pwiz_tools/Skyline/FileUI/SkypSupport.cs

```diff
@@ -156,49 +324,54 @@ public WebDownloadClient(IProgressMonitor progressMonitor, IProgressStatus progr
             ProgressStatus = progressStatus;
         }
 
-        public void Download(Uri remoteUri, string downloadPath, string username, string password)
+        public void Download(SkypFile skyp)
         {
             using (var wc = new UTF8WebClient())
             {
-                if (!string.IsNullOrEmpty(username) && !string.IsNullOrEmpty(password))
+                if (skyp.HasCredentials())
                 {
-                    wc.Headers.Add(HttpRequestHeader.Authorization, Server.GetBasicAuthHeader(username, password));
+                    wc.Headers.Add(HttpRequestHeader.Authorization, Server.GetBasicAuthHeader(skyp.Server.Username, skyp.Server.Password));
                 }
 
-                wc.DownloadProgressChanged += wc_DownloadProgressChanged;
-                wc.DownloadFileCompleted += wc_DownloadFileCompleted;
+                wc.DownloadProgressChanged += (s,e) =>
+                {
+                    // The Content-Length header is not set in the response from PanoramaWeb, so the ProgressPercentage remains 0
+                    // during the download. If the skyp includes the file size, use that to calculate progress percentage.
+                    int progressPercent = e.ProgressPercentage > 0 ? e.ProgressPercentage : -1;
+                    var fileSize = skyp.Size;
+                    if (progressPercent == -1 && fileSize.HasValue && fileSize > 0)
+                    {
+                        progressPercent = Math.Max(0, Math.Min(100, (int)(e.BytesReceived * 100 / fileSize)));
+                    }
+                    ProgressMonitor.UpdateProgress(ProgressStatus = ProgressStatus.ChangePercentComplete(progressPercent));
+                };
+
+                wc.DownloadFileCompleted += (s, e) =>
+                {
+                    if (e.Error != null && !ProgressMonitor.IsCanceled)
+                    {
+                        ProgressMonitor.UpdateProgress(ProgressStatus = ProgressStatus.ChangeErrorException(SkypDownloadException.Create(skyp, e.Error)));
+                    }
+
+                    DownloadComplete = true;
+                };
 
-                wc.DownloadFileAsync(remoteUri, downloadPath);
+                wc.DownloadFileAsync(skyp.SkylineDocUri, skyp.DownloadPath);
 
                 while (!DownloadComplete)
```

Comment:
I think something like `Thread.Sleep(100)` might be good in this loop to reduce CPU usage. Also can `DownloadComplete` be a local to the function instead of a member variable?

Reply:
Added `Thread.Sleep`. Thanks!